text/0000-io-os-reform.md

```diff
@@ -0,0 +1,1717 @@
+commit to its API.
+
+The `open_mode` function will take an `OpenOptions` struct, which will
+encompass today's `FileMode` and `FileAccess` and support a
+builder-style API.
+
+#### File kinds
+[File kinds]: #file-kinds
+
+The `FileType` module will be renamed to `FileKind`, and the
+underlying `enum` will be hidden (to allow for platform differences
+and growth). It will expose at least `is_file` and `is_dir`; the other
+methods need to be audited for compatibility across
+platforms. Platform-specific kinds will be relegated to extension
+traits in `std::os::platform`.
+
+#### File permissions
+[File permissions]: #file-permissions
+
+Unfortunately, the permission models on Unix and Windows vary
+greatly. Rather than offer an API that has no meaning on some
+platforms, we will provide a very limited `FilePermissions` structure
+in `std::fs`, and then rich extension traits in `std::os::unix` and
+`std::os::windows`.
+
+On the Unix side, the constructors and accessors for `FilePermissions`
+will resemble the flags we have today; details are left to the implementation.
+
+On the Windows side, initially there will be no extensions, as Windows
+has a very complex permissions model that will take some time to build
+out.
+
+For `std::fs` itself, `FilePermissions` will provide constructors and
+accessors for "world readable" -- and that is all. At the moment, that
+is all that is known to be compatible across the platforms that Rust
+supports.
+
+#### `PathExt`
+[PathExt]: #pathext
+
+This trait will essentially remain stay as it is (renamed from
+`PathExtensions`), following the same changes made to `fs` free functions.
+
+#### Items to move to `os::platform`
+
+* `change_file_times` will move to `os::unix` for now (cf
+  `SetFileTime` on Windows). Eventually we will add back a
+  cross-platform function, when we have grown a notion of time in
+  `std` and have a good compatibility story across all platforms.
+
+* `lstat` will move to `os::unix` since it is not yet implemented for
+  Windows.
+
+* `chown` will move to `os::unix` (it currently does *nothing* on
+  Windows), and eventually `os::windows` will grow support for
+  Windows's permission model. If at some point a reasonable
+  intersection is found, we will re-introduce a cross-platform
+  function in `std::fs`.
+
+* In general, offer all of the `stat` fields as an extension trait on
+  `FileAttr` (e.g. `os::unix::FileAttrExt`).
+
+### `std::net`
+[std::net]: #stdnet
+
+The contents of `std::io::net` submodules `tcp`, `udp`, `ip` and
+`addrinfo` will be retained but moved into a single `std::net` module;
+the other modules are being moved or removed and are described
+elsewhere.
+
+#### TCP
+[TCP]: #tcp
+
+For `TcpStream`, the changes are most easily expressed by giving the signatures directly:
+
+```rust
+// TcpStream, which contains both a reader and a writer
+
+impl TcpStream {
+    fn connect<A: ToSocketAddr>(addr: A) -> IoResult<TcpStreama>;
```

Comment:
Would it be possible at this stage to pull the connect out to a trait? Currently there's no way that I know of to test code that includes connect() without making actual network connections. That means tests for TcpStream must actually build the compatible server too in the same project. It would be great to be able to create a mock reader/writer after accept.